### PR TITLE
[FLOSS H3] docs: extend check_doc.sh

### LINKED ISSUE #4618: add consistency check for contrib
Extend ./tests/shell/check_doc.sh to also work for:

- [ ] doc/contrib
- [ ] doc/tutorials

### PR DESCRIPTION
**Pull request by joni1993**
fixes #4618

## Basics

- [x] Short descriptions of your changes are in the release notes
      (added as entry in `doc/news/_preparation_next_release.md` which
      contains `_(my name)_`)
      **Please always add something to the release notes.**
- [x] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [x] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildservers are happy. If not, fix **in this order**:
  - [ ] add a line in `doc/news/_preparation_next_release.md`
  - [ ] reformat the code with `scripts/dev/reformat-all`
  - [ ] make all unit tests pass
  - [ ] fix all memleaks
- [x] The PR is rebased with current master.

## Checklist

- [ ] I added unit tests for my code
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation (see [Documentation Guidelines](https://www.libelektra.org/devgettingstarted/documentation))
- [ ] I fixed all affected decisions (see [Decision Process](https://www.libelektra.org/decisions/decision-process))
- [ ] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://www.libelektra.org/devgettingstarted/coding))
- [ ] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [ ] I mentioned [every code](/.reuse/dep5) not directly written by me in [reuse syntax](https://reuse.software/)

## Review

- [ ] Documentation is introductory, concise, good to read and describes everything what the PR does
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

## Labels

- [ ] Add the "work in progress" label if you do not want the PR to be reviewed yet.
- [x] Add the "ready to merge" label **if the basics are fulfilled** and no further pushes are planned by you.

### CODE REVIEW

**Review by markus2330 — CHANGES_REQUESTED** [review]
Thank you, great to have this! Please fix that the script might silently ignore moved folders.

**Comment by flo91** [comment]
Please add the label "H2" if you want to use this PR for FLOSS H2.

This PR needs a restyle and a rebase.
You can apply the changes for restyling with a CI provided patch.
Just click on "[Details](https://restyled.io/gh/ElektraInitiative/repos/libelektra/jobs/2485877)" next to the line with "restyled" in the CI overview.

Otherwise, the PR looks good.

@markus2330 Please confirm the new changes.

**Review by markus2330 — APPROVED** [review]
LGTM, please rebase&restyle

**Review by markus2330 — APPROVED** [review]
LGTM

**Comment by joni1993** [comment]
It appears to me that the restyle check always fails because of the underscore in the filename. When escaping the underscore `\_` it complained it shouldn't have the escaping backslash - now with the escaping-backslash removed (which is wrong i think) it says it should be something totally wrong (`Extend check*doc.sh to work for contrib and tutorials *(@Joni1993)\_`)

**Comment by joni1993** [comment]
> It appears to me that the restyle check always fails because of the underscore in the filename. When escaping the underscore `\_` it complained it shouldn't have the escaping backslash - now with the escaping-backslash removed (which is wrong i think) it says it should be something totally wrong (`Extend check*doc.sh to work for contrib and tutorials *(@Joni1993)\_`)

I guess the restyle check is buggy here @markus2330 ?

**Comment by kodebach** [comment]
Could be a bug, yes. Although we use the widely-used `prettier` for Markdown, so there is probably a way to make it work.

File: doc/news/_preparation_next_release.md [hunk 1]
```diff
@@ -283,6 +283,7 @@ This section keeps you up-to-date with the multi-language support provided by El
 - <<TODO>>
 - Fix grammar for `elektra-granularity.md` _(@dtdirect)_
 - Rephrase sections in doc/dev/error-\* _(@dtdirect)_
+- Extend check_doc.sh to work for contrib and tutorials _(@Joni1993)_
```
Comment (kodebach) [inline doc/news/_preparation_next_release.md:286]: I believe this should work:

```suggestion
- Extend check\_doc.sh to work for contrib and tutorials _(@Joni1993)_
```

Alternatively, try this:

```suggestion
- Extend `check_doc.sh` to work for contrib and tutorials _(@Joni1993)_
```

Comment (joni1993) [reply doc/news/_preparation_next_release.md]: > - Extend check\_doc.sh to work for contrib and tutorials _(@Joni1993)_

Had this before - restyled suggested this patch: https://restyled.io/gh/ElektraInitiative/repos/libelektra/jobs/2485877/patch

**Comment by markus2330** [comment]
jenkins build libelektra please

File: doc/news/_preparation_next_release.md [hunk 2]
```diff
@@ -283,6 +283,7 @@ This section keeps you up-to-date with the multi-language support provided by El
 - <<TODO>>
 - Fix grammar for `elektra-granularity.md` _(@dtdirect)_
 - Rephrase sections in doc/dev/error-\* _(@dtdirect)_
+- Extend `check_doc.sh` to work for contrib and tutorials _(@Joni1993)_
```
Comment (markus2330) [inline doc/news/_preparation_next_release.md:286]:
```suggestion
- Extend consistency check `check_doc.sh` to work for [contrib](/doc/contrib), [dev](/doc/dev) and [tutorials](/doc/tutorials) _(@Joni1993)_
```

File: doc/Git.md [hunk 3]
```diff
@@ -0,0 +1,117 @@
+# Git
```
Comment (markus2330) [inline doc/Git.md:1]: Please make sure that `doc/GIT.md` also gets removed within the same commit so that the rename gets detected.

**Review by markus2330 — CHANGES_REQUESTED** [review]
Thank you, great work!

Renaming Git would have better fit in a separate PR, as it is unrelated to this and now two unrelated problems hinder this PR to get merged.

CI still fails because of the consistency check, please add the trivial `doc/dev/README.md` to let it succeed.

CI fails with:

```


#15 33.99 Running check_doc

#15 34.06 

#15 34.06 

#15 34.06 ELEKTRA DOCU CONSISTENCY CHECK

#15 34.06 

#15 34.06 fatal (no cleanup necessary): /home/jenkins/workspace/libelektra_PR-4634@2/libelektra/doc/dev/README.md not found!

#15 34.06 error: check_doc
```

https://build.libelektra.org/blue/organizations/jenkins/libelektra/detail/PR-4634/11/pipeline

File: doc/news/_preparation_next_release.md (hunk 1 above)
Comment (markus2330) [reply doc/news/_preparation_next_release.md]: Discussion moved to: #4687

**Comment by joni1993** [comment]
> Thank you, great work!
> 
> Renaming Git would have better fit in a separate PR, as it is unrelated to this and now two unrelated problems hinder this PR to get merged.
> 
> CI still fails because of the consistency check, please add the trivial `doc/dev/README.md` to let it succeed.
> 
> CI fails with:
> 
> ```
> 
> 
> #15 33.99 Running check_doc
> 
> #15 34.06 
> 
> #15 34.06 
> 
> #15 34.06 ELEKTRA DOCU CONSISTENCY CHECK
> 
> #15 34.06 
> 
> #15 34.06 fatal (no cleanup necessary): /home/jenkins/workspace/libelektra_PR-4634@2/libelektra/doc/dev/README.md not found!
> 
> #15 34.06 error: check_doc
> ```
> 
> https://build.libelektra.org/blue/organizations/jenkins/libelektra/detail/PR-4634/11/pipeline

I don't quiet get why this is. The file is there like 6 years (https://github.com/joni1993/libelektra/blame/checkdocsh/doc/dev/README.md). But i have to say i am unsure how the file is actually called during the test process.

**Comment by kodebach** [comment]
> But i have to say i am unsure how the file is actually called during the test process.

AFAICT in this case, `kdb run_all` is executed in a Docker container using one of the `release.Dockerfile` images, e.g. for [Debian bullseye](https://github.com/ElektraInitiative/libelektra/blob/eb77b54f3d45b45f3b17fa61c7501e8fe5b4d9b3/scripts/docker/debian/bullseye/release.Dockerfile). From what I see, this image first installs a packaged version of Elektra, so probably the file is not included in our packages.

I'm not sure the file _should_ be part of the packages. @markus2330 should know. If it should be in the packages, you need an `install` with `COMPONENT` in a CMake file somewhere. If the file should not be in the packages, then the `testscr_check_doc` probably shouldn't run for installed version (i.e. not run with `kdb run_all`). For that you need to add the test to the `excluded_tests` in `tests/shell/CMakeLists.txt`.

**Comment by markus2330** [comment]
I agree, check_doc doesn't make much sense on an installed system. Exclusion is possible in tests/shell/CMakeLists.txt lines 57:

https://github.com/ElektraInitiative/libelektra/blob/f7874ef5a3fac57dc157fb882e6ee818e6c759cc/tests/shell/CMakeLists.txt#L57-L65

**Comment by joni1993** [comment]
> You added the "H3"-Label to this PR.
> Are you sure that you want to use this PR for **H3 instead of H2?**
> In the preliminary grading, I thought this PR was intended for H2, because it had no label assigned or prefix in the title.
> If you don't use this PR for H2 and I didn't miss something, you've only 4 other PRs for H2 submitted.

This is fine.

**Review by markus2330 — APPROVED** [review]

**Comment by markus2330** [comment]
The changes are fine. The only left-over is to fix the tests by excluding the tests, see https://github.com/ElektraInitiative/libelektra/pull/4634#issuecomment-1326277330

File: doc/news/_preparation_next_release.md [hunk 4]
```diff
@@ -338,7 +338,7 @@ This section keeps you up-to-date with the multi-language support provided by El
 - Fix grammar for `elektra-granularity.md` _(@dtdirect)_
 - Rephrase sections in doc/dev/error-\* _(@dtdirect)_
 - Unify spelling of man pages _(@stefnotch)_ _(@janldeboer)_
-- <<TODO>>
+- Extend consistency check `check_doc.sh` to work for [contrib](/doc/contrib), [dev](/doc/dev) and [tutorials](/doc/tutorials) _(@Joni1993)
```
Comment (flo91) [inline doc/news/_preparation_next_release.md:341]: The `_` at the end of the line was missing.

```suggestion
- Extend consistency check `check_doc.sh` to work for [contrib](/doc/contrib), [dev](/doc/dev) and [tutorials](/doc/tutorials) _(@Joni1993)_
```

**Comment by markus2330** [comment]
Please also rebase (take your changes to doc/contrib/README.md).

**Comment by joni1993** [comment]
> The problem is still not fixed, please exclude test on installed system as adviced in: [#4634 (comment)](https://github.com/ElektraInitiative/libelektra/pull/4634#issuecomment-1326277330)

Sorry i don't know what has to be done further than this:
https://github.com/ElektraInitiative/libelektra/pull/4634/commits/dc08e8b11d6e25d00f359a20508a9c32894f6e50#diff-a1edda73985c488a09daede8c20dc0cd0f52b116703bb05220b1b4f08b5b4535R60

**Comment by joni1993** [comment]
jenkins build libelektra please

**Comment by joni1993** [comment]
jenkins build libelektra please one more time

**Comment by joni1993** [comment]
jenkins build libelektra please

**Comment by joni1993** [comment]
After recent Jenkins hickups - it is happy now. Maybe @markus2330 can merge this soon before doc changes one more time?

**Comment by flo91** [comment]
It's merged now. 😃 
Thank you and Happy Holidays! 🎁 🎆

**Pull request merged**